### PR TITLE
  Add native bluetooth BLE support

### PR DESCRIPTION
**Pull request by AdrianSoundy**
## Description

Support added for Nimble Bluetooth BLE support to work with new nanoFramework.Device.Bluetooth
Initial version. 

This supports the GATT server services and no client support

Support for
- GATT Service.
- GATT Characteristics & User and Presentation format descriptors.
- GATT Characteristics general descriptors.
- Handling of Read Characteristics with managed events and static values.
- Handling of Write/Write no repsonse Characteristics.
- Handling of Notify/Indicate.
- Events to managed code for Read/Write/Notify.Client subscribe.
- Advertising Discoverable / Connectable / Device name.

Other changes
- Removed existing bluetooth test code and now renamed to nanoFramework.Device.Bluetooth.
- Reworked Bluetooth makelist.txt.
- Updated Azure-pipelines builds.
- Added new Sdk config(sdkconfig.default_nopsram_ble.esp32) for Rev0/Ble with no spiram otherwise it doesn't fit in iram. The work-around fixes for Spiram cause extra 10K to be added to iram(Newlib patches) if using spiram.
- Current reduce managed heap for Rev0/Ble/no spiram config (further testing on amount o reduce required).

Notes:
The current IDF debug builds don't include symbols and seem to be same as MinSizeRel builds. MinSizeRel doesn't seem to change anything in Cmake project.
The Rev3/Ble can't be built in proper debug mode as IRAM overflow. RTM just fits (99.2%). Debugging was done with Rev0/Ble build.

## How Has This Been Tested?

- Using some test cases with Android apps "LightBlue", "BLE scanner", nRf Connect & Serial Bluetooth Terminal 
- Tests & samples will be added as part of managed code

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Unit Tests (work on Unit Tests, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/esp32_nimble.cpp [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+                "connection %s; status=%d\n",
+                event->connect.status == 0 ? "established" : "failed",
+                event->connect.status);
+
+            break;
+
+        case BLE_GAP_EVENT_DISCONNECT:
+            ESP_LOGI(tag, "BLE_GAP_EVENT_DISCONNECT; reason=%d\n", event->disconnect.reason);
+
+            /* Connection terminated; resume advertising */
```
Comment (josesimoes) [inline targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/esp32_nimble.cpp:80]:
```suggestion
            // Connection terminated; resume advertising
```

File: targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/esp32_nimble.cpp [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+{
+    ESP_LOGI(tag, "BLE on reset %d\n", reason);
+}
+
+void esp32_ble_host_task(void *param)
+{
+    (void)param;
+
+    ESP_LOGI(tag, "Bluetooth LE Host Task Started");
+    /* This function will return only when nimble_port_stop() is executed */
```
Comment (josesimoes) [inline targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/esp32_nimble.cpp:277]:
```suggestion
    // This function will return only when nimble_port_stop() is execute
```

File: targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_dev_ble_native_nanoFramework_Device_Bluetooth_GenericAttributeProfile_GattServiceProvider.cpp [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+        (CLR_RT_HeapBlock_Array *)pGattLocalService[GattLocalService::FIELD___serviceUuid].Array();
+
+    pUuid = pUuidArray->GetFirstElement();
+
+    CLR_RT_HeapBlock_ArrayList *pCharacteristics =
+        (CLR_RT_HeapBlock_ArrayList *)pGattLocalService[GattLocalService::FIELD___characteristics].Dereference();
+
+    CLR_INT32 CharacteristicsCount = pCharacteristics->GetSize();
+
+    CharacteristicsCount++; // +1 for Terminator definition
```
Comment (josesimoes) [inline targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_dev_ble_native_nanoFramework_Device_Bluetooth_GenericAttributeProfile_GattServiceProvider.cpp:276]:
```suggestion

    // +1 for Terminator definition
    CharacteristicsCount++;
```

File: targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_dev_ble_native_nanoFramework_Device_Bluetooth_GenericAttributeProfile_GattLocalCharacteristic.cpp [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+{
+    NANOCLR_HEADER();
+    {
+        uint16_t characteristicId;
+        uint16_t connectionHandle;
+        uint16_t attHandle;
+        CLR_RT_HeapBlock_Array *notifyBuffer;
+        uint8_t *bufPtr;
+        int bufLen;
+        int rc = 1; // error characteristicId not known
```
Comment (josesimoes) [inline targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_dev_ble_native_nanoFramework_Device_Bluetooth_GenericAttributeProfile_GattLocalCharacteristic.cpp:28]:
```suggestion
        // error characteristicId not known
        int rc = 1;
```

File: targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_ble_nimble_chars.c [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+static int gatt_svr_chr_access_device_info(
+    uint16_t conn_handle,
+    uint16_t attr_handle,
+    struct ble_gatt_access_ctxt *ctxt,
+    void *arg);
+
+const struct ble_gatt_chr_def gatt_char_device_info[] = {
+    {
+        //* Characteristic:  Manufacturer name
```
Comment (josesimoes) [inline targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_ble_nimble_chars.c:35]:
```suggestion
        // Characteristic:  Manufacturer name
```

File: targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_ble_nimble_chars.c [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        .flags = BLE_GATT_CHR_F_READ,
+    },
+    {
+        // Characteristic: Model number string
+        .uuid = BLE_UUID16_DECLARE(GATT_MODEL_NUMBER_UUID),
+        .access_cb = gatt_svr_chr_access_device_info,
+        .flags = BLE_GATT_CHR_F_READ,
+    },
+    {
+        0, /* No more characteristics in this service */
```
Comment (josesimoes) [inline targets/ESP32/_nanoCLR/nanoFramework.Device.Bluetooth/sys_ble_nimble_chars.c:47]:
```suggestion
        // No more characteristics in this service
        0,
```

**Review by josesimoes — COMMENTED** [review]
Nice!! Looking good! 👏🏻

**Comment by AdrianSoundy** [comment]
@josesimoes 
Would like to get this merged now. Are you ok with this ?

**Review by josesimoes — APPROVED** [review]
LGTM! Ok for the merge.🙂👍

The stubs for the win32 CLR are missing, but can be added later.

**Pull request merged**